File: gunkim/leetcode/array/152.Maximum Product Subarray.java
public class Main {
    public static void main(String[] args) {
        int[] testArr = {0,2};
        Solution solution = new Solution();
        System.out.println(solution.maxProduct(testArr));
    }
}
class Solution {
    public int maxProduct(int[] nums) {
        if(nums.length == 1) {
            return nums[0];
        }
        int max = Integer.MIN_VALUE;
        for(int i = 0; i < nums.length; i++) {
            if(nums[i] == 0) continue;
            if(nums[i] > max) max = nums[i];

            int currentMax = nums[i];
            for(int j = (i+1); j < nums.length; j++) {
                currentMax *= nums[j];

                if(currentMax > max) {
                    max = currentMax;
                }
            }
            if(nums[nums.length - 1] > max) max = nums[nums.length - 1];
        }
        return max;
    }
}

File: anb/leetcode/array/152. Maximum Product Subarray.cpp
const int INF = (1LL << 31) - 1;
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ans = -INF;
        pair<int, int> bef(1,1), cur;
        for(int& n : nums){
            cur.first = max({n, n * bef.first, n * bef.second});
            cur.second = min({n, n * bef.first, n * bef.second});
            
            bef = cur;            
            ans = max({ans, n, cur.first});
        }
        return ans;
    }
};